File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n+1,0);
        dp[0] = nums[0];
        for(int i=1;i<n;i++){
            int pick = nums[i];
            if(i>1){
                pick += dp[i-2];
            }
            int nonpick = dp[i-1];
            cout<<pick<<" "<<nonpick<<endl;
            dp[i] = max(pick,nonpick);
        }
        return dp[n-1];
    }
};

File: 0198-house-robber/0198-house-robber.ts
function rob(nums: number[]): number {
    if (nums.length === 0) return 0; // Handle empty array case
    const dp: number[] = new Array(nums.length).fill(-1);
    return rec(nums.length - 1, nums, dp);
}

function rec(i: number, nums: number[], dp: number[]): number {
    if (i < 0) return 0; // Base case: No house to rob
    if (dp[i] !== -1) return dp[i]; // Return cached value

    dp[i] = Math.max(rec(i - 1, nums, dp), nums[i] + rec(i - 2, nums, dp));
    return dp[i];
}
